core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -487,19 +505,25 @@ class LogSegment(val log: FileRecords,
   /**
    * Delete this log segment from the filesystem.
    */
-  def delete() {
-    val deletedLog = log.delete()
-    val deletedIndex = index.delete()
-    val deletedTimeIndex = timeIndex.delete()
-    val deletedTxnIndex = txnIndex.delete()
-    if (!deletedLog && log.file.exists)
-      throw new IOException("Delete of log " + log.file.getName + " failed.")
-    if (!deletedIndex && index.file.exists)
-      throw new IOException("Delete of index " + index.file.getName + " failed.")
-    if (!deletedTimeIndex && timeIndex.file.exists)
-      throw new IOException("Delete of time index " + timeIndex.file.getName + " failed.")
-    if (!deletedTxnIndex && txnIndex.file.exists)
-      throw new IOException("Delete of transaction index " + txnIndex.file.getName + " failed.")
+  def deleteIfExists() {
+    def delete(delete: () => Boolean, fileType: String, file: File, logIfMissing: Boolean): Unit = {
+      try {
+        if (delete())
+          info(s"Deleted $fileType ${file.getAbsolutePath}.")
+        else if (logIfMissing)
+          info(s"Failed to delete $fileType ${file.getAbsolutePath} because it does not exist.")
+      }
+      catch {
+        case e: IOException => throw new IOException(s"Delete of $fileType ${file.getAbsolutePath} failed.", e)
+      }
+    }
+
+    CoreUtils.tryAll(Seq(
+      () => delete(log.deleteIfExists _, "log", log.file, logIfMissing = true),
+      () => delete(offsetIndex.deleteIfExists _, "offset index", offsetIndex.file, logIfMissing = true),
+      () => delete(timeIndex.deleteIfExists _, "time index", timeIndex.file, logIfMissing = true),
+      () => delete(txnIndex.deleteIfExists _, "transaction index", txnIndex.file, logIfMissing = false)
```

Comment:
Interesting. Why would we not want to log the missing file for txnIndex?

Reply:
My understanding is that the transactions index is only created if there are aborted transactions. For the majority of cases where people are not using transactions, this would be noise. Or am I missing something?

Reply:
Thanks for the explanation. That makes sense. The timeindex may not be present during the upgrade. Since that's rare, it's probably ok.

Reply:
Yes, that was my thought as well.